open-craft/problem-builder
# Step Builder: Add mentoring-level messages to review step and disable mode switching for Problem Builder

This PR adds three types of customizable messages to review step blocks:

* **Complete** - shown if all answers correct (and some attempts remaining)
* **Incomplete** - shown if one or more answers are incorrect or partially correct (and some attempts remaining)
* **Review** - shown when user ran out of attempts

As a first step towards deprecating assessment functionality of Problem Builder, this PR removes the ability to change the mode of Problem Builder blocks to assessment mode.

Additionally, it includes the following fixes/additions:

* Sequence navigation tooltips do not display Step Builder descendants anymore.
* Step Builder sends a `"grade"` event when user completes an attempt (submits the last step).

**Testing**

1. If you haven't already, enable Step Builder by adding `"step-builder"` to advanced modules.
2. Add Step Builder to a new or existing unit and set max attempts to some non-zero value.
3. Add some Mentoring Step blocks and a Review Step block.
4. Add some questions to each Mentoring Step block.
5. Add message blocks to Review Step block, optionally customizing their content. Observe that each type of message can only be added once.
6. Complete the Step Builder block in the LMS. Observe that review step will display "Complete" message if all answers are correct, "Incomplete" message if some answers are incorrect/partially correct, "Review" message when no attempts remain.
